art/attacks/poisoning/gradient_matching_attack.py

```diff
@@ -0,0 +1,556 @@
+            if B_ < best_B:
+                best_B = B_  # pylint: disable=C0103
+                best_x_poisoned = x_poisoned
+                best_indices_poison = indices_poison
+
+        x_train[best_indices_poison] = best_x_poisoned
+        return x_train, y_train  # y_train has not been modified.
+
+    def __poison__pytorch(
+        self, x_trigger: np.ndarray, y_trigger: np.ndarray, x_poison: np.ndarray, y_poison: np.ndarray  # pylint: disable=unused-argument
+    ) -> np.ndarray:
+        """
+        Optimize the poison by matching the gradient within the perturbation budget.
+
+        :param x_trigger: List of triggers.
+        :param y_trigger: List of target labels.
+        :param x_poison: List of samples to poison.
+        :param y_poison: List of the labels for x_poison.
+        :return: A pair of poisoned samples, B-score (cosine similarity of the gradients).
+        """
+
+        import torch
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        class PoisonDataset(torch.utils.data.Dataset):
+            """
+            Iterator for a dataset to poison.
+            """
+
+            def __init__(self, x: np.ndarray, y: np.ndarray):
+                self.len = x.shape[0]
+                self.x = torch.as_tensor(x, device=device, dtype=torch.float)
+                self.y = torch.as_tensor(y, device=device, dtype=torch.long)
+
+            def __getitem__(self, index):
+                return self.x[index], torch.as_tensor([index], device=device), self.y[index]
+
+            def __len__(self):
+                return self.len
+
+        trainloader = torch.utils.data.DataLoader(
+            PoisonDataset(x_poison, y_poison), batch_size=self.batch_size, shuffle=True, num_workers=8
+        )
+
+        epoch_iterator = trange(self.max_epochs) if self.verbose else range(self.max_epochs)
+        for _ in epoch_iterator:
+            batch_iterator = tqdm(trainloader) if self.verbose else trainloader
+            for x, indices, y in batch_iterator:
+                self.backdoor_model.zero_grad()
+                loss, poisoned_samples = self.backdoor_model(x, indices, y, self.grad_ws_norm)
+                self.backdoor_model.noise_embedding.embedding_layer.weight.grad.sign_()
+                loss.backward()
+                self.optimizer.step()
+            self.lr_schedule.step()
+
+        B, poisoned_samples = self.backdoor_model(  # pylint: disable=C0103
+            torch.tensor(x_poison, device=device, dtype=torch.float),
+            torch.arange(0, len(x_poison), dtype=torch.int32),
+            torch.tensor(y_poison, device=device, dtype=torch.long),
+            self.grad_ws_norm,
+        )
+        return B.detach().numpy(), poisoned_samples.detach().numpy()
+
+    def __poison__tensorflow(
+        self, x_trigger: np.ndarray, y_trigger: np.ndarray, x_poison: np.ndarray, y_poison: np.ndarray  # pylint: disable=unused-argument
```

Comment:
Why is there an unused argument?

Reply:
Fixed. It is a remnant from the previous method definition when it was overriding poison.